src/main/java/com/bakdata/kafka/S3BackingClient.java

```diff
@@ -0,0 +1,91 @@
+package com.bakdata.kafka;
+
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3URI;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.amazonaws.services.s3.model.S3Object;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.Objects;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.common.errors.SerializationException;
+
+@Slf4j
+@RequiredArgsConstructor
+class S3BackingClient {
```

Comment:
Even though this class is not public, it would be helpful to have JavaDocs or at least a comment explaining how we serialize records, e.g., that the first byte is a flag indicating whether the record has been backed or not.